### ISSUE TITLE
fmpz_mat view and aliasing in fmpz_mat_mul not working properly

### ISSUE THREAD

**Issue by thofma** [wbhart/flint2#891]
```julia
julia> M = ZZ[2;]
[2]

julia> typeof(M)
fmpz_mat

julia> MM = view(M, 1:1, 1:1)
[2]

julia> mul!(MM, MM, MM)
[4]

julia> MM
[4]

julia> M
[2]
```
I think the matrix `M` should also have been changed. It works for example with `mul!(MM, ZZ[2;], ZZ[2;])`.

**Comment by tthsqe12** [comment]
This also corrupts memory right?

The issue should be these two lines
https://github.com/wbhart/flint2/blob/trunk/fmpz_mat/mul.c#L293#L292

The mul function should not be clearing it's input because it doesn't know if it is a window (in which case it should be using `fmpz_mat_window_clear`.

**Comment by thofma** [comment]
Yes, it is also corrupting memory, but it is a bit harder to reproduce and I don't have a nice MWE.

**Comment by tthsqe12** [comment]
That is fine. The problem is clear enough. I see two solutions:
- use set&clear so that thing being cleared on line 293 is definitely not a view.
- introduce a flag for windows so that views and non-views have the same clear function.

**Comment by thofma** [comment]
Someone is trying to fix it for arb: https://github.com/fredrik-johansson/arb/pull/362

**Comment by fredrik-johansson** [comment]
Anyone here have any thoughts on the proposed fix in fredrik-johansson/arb#362?

We could also document explicitly that aliased multiplication of window matrices isn't legal. You'll run into trouble with partially overlapping windows anyway, and this limitation probably affects other operations as well. But perhaps multiplication is useful enough that we should handle it as a convenience.

**Comment by thofma** [comment]
(FWIW, I find it very useful.)

**Comment by wbhart** [comment]
In all the examples there, this only affects aliased matrices, as the swapping is done to handle that. I think we either document that this is not supported, or we do the entry-wise copy. I'm ok with either solution.

From memory we don't support aliasing of matrices in general. Is that correct?

Will we meet again tomorrow for Flint-Fridays? We could discuss this there.

**Comment by fredrik-johansson** [comment]
> From memory we don't support aliasing of matrices in general. Is that correct?

We didn't originally but aliasing should work in most places now, so we may as well fix any remaining cases we find that are broken.

> Will we meet again tomorrow for Flint-Fridays? We could discuss this there.

Yes.

**Comment by wbhart** [comment]
This doesn't work. See the comments on the PR https://github.com/wbhart/flint2/pull/940

**Comment by wbhart** [comment]
Fixed by #940
